<h1 style='text-align : center '>Sets</h1>

<h2>4 important rules of sets :</h2>
<ol>
    <li>Sets do not allow duplicates.</li>
    <li>Sets have no indexing/slicing.</li>
    <li>Sets don't allow mutable data types in the sets.</li>
    <li>Set itself is a mutable data type.</li>
</ol>

<br>
<h2>Create</h2>

In [1]:
# want to create empty sets
s1={} # this is not set
s1

{}

In [2]:
type(s1) # s1 is dictionary not set

dict

In [3]:
s2=set() # s2 is  set
s2

set()

In [4]:
type(s2)

set

In [5]:
s1={1,2,3,4,5}
s1

{1, 2, 3, 4, 5}

In [6]:
s2={"Hello",True,5,4.5}
s2

{4.5, 5, 'Hello', True}

In [7]:
s3={1,1,2,2,3,3} #Sets do not allow duplicates.
s3

{1, 2, 3}

In [8]:
s4={{1,2,3,4},"Hello"} # Sets don't allow mutable data types in the sets but Set itself is a mutable data type for that reason get an error
s4

TypeError: unhashable type: 'set'

In [11]:
s5={(1,2,3),"Hello",4} # tuples (1,2,3,4) are not mutable
s5   #sets have no indexing, Sets follow Hashing

{(1, 2, 3), 4, 'Hello'}

In [12]:
s6={{1,2},{3,4}} #Set itself is a mutable data type.
s6

TypeError: unhashable type: 'set'

<br>
<h2>Access items</h2>

In [13]:
s3={1,2,3,4,5} #Sets have no indexing.
s3[0]

TypeError: 'set' object is not subscriptable

In [14]:
s3[0:4] #Sets have no slicing.

TypeError: 'set' object is not subscriptable

<br>
<h2>Edit items</h2>

In [15]:
s3={1,2,3,4,5} # we can not edit items
s3[0] = 100
s3

TypeError: 'set' object does not support item assignment

<h1>If we change sets to list, then edit list after that list to sets. this way will working but this is not editing </h1>

In [18]:
s1={1,2,3,4,5}
id(s1) # for memorry address

1942229537216

In [19]:
s2=list(s1)
s2[0]=100
s3=set(s2)
s3

{2, 3, 4, 5, 100}

In [21]:
id(s3)

1942229536992

<h2> s1 memory address is 1942229537216 and s3=1942229536992,memory address are not same . we can say if this way did edit in sets then the memory address would be the same</h2>

<br>
<h2>Add items</h2>

In [24]:
s1={1,2,3,4,90}
s1

{1, 2, 3, 4, 90}

In [25]:
id(s1)

1942229538336

In [26]:
s1.add(20)

In [27]:
s1

{1, 2, 3, 4, 20, 90}

In [29]:
id(s1)

1942229538336

<h2>Delete items :</h2>
<ul>
    <li>del</li>
    <li>remove</li>
    <li>pop</li>
</ul>

In [30]:
s2

[100, 2, 3, 4, 5]

In [31]:
del s2

In [32]:
s2

NameError: name 's2' is not defined

In [33]:
s1

{1, 2, 3, 4, 20, 90}

In [34]:
s1.remove(90)
s1

{1, 2, 3, 4, 20}

In [36]:
s1

{1, 2, 3, 4, 20}

In [37]:
s1.pop() #delete 1 bcz of Hashing

1

<br>
<h2>Operations</h2>

In [38]:
s1={1,2,3,4} # concate not support
s2={5,6,7,8}
s1+s2

TypeError: unsupported operand type(s) for +: 'set' and 'set'

In [39]:
s1*3 #multiply not support

TypeError: unsupported operand type(s) for *: 'set' and 'int'

In [40]:
for i in s1:  #loops support
    print(i)

1
2
3
4


In [41]:
1 in s1 # membership allow

True

In [42]:
5 in s1

False

<br>
<h2>Functions</h2>

In [44]:
len(s1)

4

In [45]:
max(s1)

4

In [46]:
min(s1)

1

In [47]:
sorted(s1) #change to list

[1, 2, 3, 4]

In [49]:
sorted(s1,reverse=True)

[4, 3, 2, 1]

In [50]:
s1={1,2,3,4,5}
s2={4,5,6,7,8}
s1.union(s2)

{1, 2, 3, 4, 5, 6, 7, 8}

In [51]:
s1.intersection(s2)

{4, 5}

In [52]:
s1.difference(s2)

{1, 2, 3}

In [54]:
s2.difference(s1)

{6, 7, 8}

In [55]:
s1.symmetric_difference(s2) # unique items from every sets

{1, 2, 3, 6, 7, 8}

In [56]:
s1.isdisjoint(s2) # if two sets don't have any common items = True

False

In [57]:
s1.issubset(s2) # s1 all value in s2 and s2 has some other elements = True

False

In [59]:
s1.issubset(s2) # s2 all value in s1 and s1 has some other elements = True

False

In [61]:
s1.issuperset(s2)

False